torch/csrc/jit/custom_operator.h

```diff
@@ -0,0 +1,141 @@
+}
+
+/// Create a single-element `vector` for simple (non-tuple) return types.
+template <typename ReturnType>
+std::vector<Argument> createReturns(ReturnType*) {
+  return {Argument("_1", getTypePtr<ReturnType>())};
+}
+
+/// Creates a vector of `Argument` from `FunctionTraits` and a pack of indices
+/// into the argument list.
+template <typename FunctionTraits, size_t... Is>
+std::vector<Argument> createArgumentVectorFromTraits(Indices<Is...> indices) {
+  using ArgumentTypes = typename FunctionTraits::parameter_types;
+  return createArgumentVectorFromTypes<
+      c10::guts::typelist::element_t<Is, ArgumentTypes>...>(indices);
+}
+
+/// Creates a `FunctionSchema` object from a `FunctionTraits` type for a
+/// function.
+template <typename FunctionTraits>
+FunctionSchema createFunctionSchemaFromTraits(const std::string& name) {
+  using ReturnType = typename FunctionTraits::return_type;
+  auto arguments = createArgumentVectorFromTraits<FunctionTraits>(
+      typename MakeIndices<FunctionTraits::number_of_parameters>::indices{});
+  auto returns = createReturns(static_cast<ReturnType*>(nullptr));
+  return {name, arguments, returns};
+}
+
+/// Does two things for an operator implementation and a tuple of arguments:
+/// 1. Pops all necessary arguments off the stack into the tuple's elements,
+/// 2. Unpacks the tuple and calls the operator implementation.
+/// The result of the implementation call is returned.
+template <
+    typename ReturnType,
+    typename Implementation,
+    typename... Types,
+    size_t... Is>
+ReturnType callOperatorWithTuple(
+    Implementation&& implementation,
+    Stack& stack,
+    std::tuple<Types...>& tuple,
+    Indices<Is...>) {
+  pop(stack, std::get<Is>(tuple)...);
+  return std::forward<Implementation>(implementation)(std::get<Is>(tuple)...);
+}
+} // namespace detail
+
+/// Low-level interface to register an operator with a parsed `FunctionSchema`
+/// and a stack-based operator implementation (the `operation`). The `operation`
+/// must pop its arguments from the stack, perform some operation on those
+/// arguments, and then push the return value back onto the stack.
+inline void registerOperatorWithStack(
+    FunctionSchema schema,
+    Operation operation) {
+  registerOperator({schema, [operation](Node*) { return operation; }});
+}
+
+/// Registers a custom operator with a schema and an implementation function.
+/// The implementation function can be a function pointer or a functor
+/// (including a lambda object). The function (or `operator()`) can take any
+/// number of arguments with a type from the subset accepted by the PyTorch
+/// JIT/Script backend, and return a single type or a tuple of types.
+/// Example invocation:
+/// ```
+/// registerOperator(
+///    parseSchema("foo::bar(float a, Tensor b)"),
+///    [](float a, at::Tensor b) { return a + b; });
+/// ```
+template <typename Implementation>
+void registerOperator(FunctionSchema schema, Implementation&& implementation) {
```

Comment:
It would be nice if these functions just returned Operator so that they can be used with the `RegisterOperators` class. Easy change.

Reply:
Also, can we merge this one with the the `std::string` variant? You can tell if you have a full declaration or just a name by the presence of '('. That would avoid having the user need to call parseSchema, and allow us to delay schema parsing past static-initialization time, which is important for debugging.